File: android-example/app/src/main/java/com/dsrv/wallet/example/wallet/model/WalletHandlers.kt
package com.dsrv.wallet.example.wallet.model

import com.dsrv.wallet.sdk.AuthHandler
import com.dsrv.wallet.sdk.ChallengeRequest
import com.dsrv.wallet.sdk.ChallengeResult
import kotlinx.coroutines.suspendCancellableCoroutine
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

/**
 * AuthHandler implementation that delegates authentication to customer backend.
 */
class MyAuthHandler(
    private val backendUrl: String
) : AuthHandler {

    private val httpClient = OkHttpClient()

    override suspend fun requestChallenge(request: ChallengeRequest): ChallengeResult {
        val jsonBody = JSONObject().apply {
            put("sdkId", request.sdkId)
            put("appId", request.appId)
            put("userCredential", JSONObject().apply {
                put("type", request.userCredential.type.name)
                put("value", request.userCredential.value)
                put("provider", request.userCredential.provider)
            })
            put("signingHash", request.signingHash)
            put("deviceInfo", JSONObject().apply {
                put("platform", request.deviceInfo.platform)
                put("publicKey", request.deviceInfo.publicKey)
                put("model", request.deviceInfo.model)
                put("osVersion", request.deviceInfo.osVersion)
                put("isVirtual", request.deviceInfo.isVirtual)
            })
        }.toString()

        return try {
            val (statusCode, responseBody) = post("$backendUrl/sdk/registration", jsonBody)
            android.util.Log.d("MyAuthHandler", "registration response [$statusCode]: $responseBody")
            val json = JSONObject(responseBody)

            val data = if (statusCode in 200..299) json.optJSONObject("data") else null
            val challenge = data?.optString("challenge")?.takeIf { it.isNotEmpty() }
            if (challenge != null) {
                ChallengeResult.Success(challenge)
            } else {
                ChallengeResult.Failure(json.optString("message", "Challenge request failed"))
            }
        } catch (e: Exception) {
            ChallengeResult.Failure(e.message ?: "Network error")
        }
    }

    private suspend fun post(url: String, jsonBody: String): Pair<Int, String> =
        suspendCancellableCoroutine { continuation ->
            val request = Request.Builder()
                .url(url)
                .post(jsonBody.toRequestBody("application/json".toMediaType()))
                .build()

            val call = httpClient.newCall(request)
            continuation.invokeOnCancellation { call.cancel() }

            call.enqueue(object : Callback {
                override fun onFailure(call: Call, e: IOException) {
                    if (continuation.isActive) continuation.resumeWithException(e)
                }

                override fun onResponse(call: Call, response: Response) {
                    if (!continuation.isActive) return
                    response.use { resp ->
                        continuation.resume(resp.code to (resp.body?.string().orEmpty()))
                    }
                }
            })
        }
}

File: ios-example/dsrv-wallet-sdk-ios-example/Models/WalletHandlers.swift
//
//  WalletHandlers.swift
//  dsrv-wallet-sdk-ios-example
//
//  AuthHandler implementation for customer backend communication.
//  Follows the Delegate pattern used by Fireblocks, Web3Auth, Cobo, etc.
//

import Foundation
import dsrv_wallet_sdk_ios

/// AuthHandler implementation that delegates authentication to customer backend.
/// Follows the Delegate pattern - SDK calls these methods, app handles backend communication.
class MyAuthHandler: AuthHandler {
    private let backendUrl: String

    init(backendUrl: String) {
        self.backendUrl = backendUrl
    }

    func requestChallenge(request: ChallengeRequest) async -> ChallengeResult {
        print("MyAuthHandler: requestChallenge(sdkId: \(request.sdkId), appId: \(request.appId)) - POST \(backendUrl)/sdk/registration")
        do {
            let body: [String: Any] = [
                "sdkId": request.sdkId,
                "appId": request.appId,
                "userCredential": [
                    "type": request.userCredential.type.rawValue,
                    "value": request.userCredential.value,
                    "provider": request.userCredential.provider
                ],
                "deviceInfo": [
                    "platform": request.deviceInfo.platform,
                    "publicKey": request.deviceInfo.publicKey as Any,
                    "model": request.deviceInfo.model,
                    "osVersion": request.deviceInfo.osVersion,
                    "isVirtual": request.deviceInfo.isVirtual,
                    "attestationObject": request.deviceInfo.attestationObject as Any
                ]
            ]

            let response = try await HttpHelper.post(
                url: "\(backendUrl)/sdk/registration",
                jsonBody: body
            )
            print("MyAuthHandler: registration response: \(response)")

            guard let data = response.data(using: .utf8),
                  let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] else {
                print("MyAuthHandler: requestChallenge() - Failed to parse JSON response")
                return .failure(error: "Failed to parse response")
            }

            guard let payload = json["data"] as? [String: Any],
                  let challenge = payload["challenge"] as? String, !challenge.isEmpty else {
                print("MyAuthHandler: requestChallenge() - Challenge is empty")
                return .failure(error: "Challenge request failed")
            }

            print("MyAuthHandler: requestChallenge() - Success, challenge: \(challenge)")
            return .success(challenge: challenge)
        } catch let error as HttpHelper.HttpError {
            switch error {
            case .httpError(let statusCode, let body):
                print("MyAuthHandler: requestChallenge() - HTTP \(statusCode): \(body)")
                return .failure(error: "HTTP \(statusCode): \(body)")
            default:
                print("MyAuthHandler: requestChallenge() - Error: \(error)")
                return .failure(error: error.localizedDescription)
            }
        } catch {
            print("MyAuthHandler: requestChallenge() - Error: \(error)")
            return .failure(error: error.localizedDescription)
        }
    }
}
